components/heap/third_party/dartino/cmpctmalloc.c

```diff
@@ -0,0 +1,938 @@
+// Copyright (c) 2016, the Dartino project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE.md file.
+//
+// Copyright (C) 2019 Toitware ApS. All rights reserved.
+
+#include "cmpctmalloc.h"
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "../../multi_heap_platform.h"
+
+typedef struct multi_heap_info cmpct_heap_t;
+
+void *multi_heap_malloc(cmpct_heap_t *heap, size_t size)
+    __attribute__((alias("cmpct_malloc_impl")));
+
+void multi_heap_free(cmpct_heap_t *heap, void *p)
+    __attribute__((alias("cmpct_free_impl")));
+
+void *multi_heap_realloc(cmpct_heap_t *heap, void *p, size_t size)
+    __attribute__((alias("cmpct_realloc_impl")));
+
+size_t multi_heap_get_allocated_size(cmpct_heap_t *heap, void *p)
+    __attribute__((alias("cmpct_get_allocated_size_impl")));
+
+cmpct_heap_t *multi_heap_register(void *start, size_t size)
+    __attribute__((alias("cmpct_register_impl")));
+
+void multi_heap_get_info(cmpct_heap_t *heap, multi_heap_info_t *info)
+    __attribute__((alias("cmpct_get_info_impl")));
+
+size_t multi_heap_free_size(cmpct_heap_t *heap)
+    __attribute__((alias("cmpct_free_size_impl")));
+
+size_t multi_heap_minimum_free_size(cmpct_heap_t *heap)
+    __attribute__((alias("cmpct_minimum_free_size_impl")));
+
+typedef uintptr_t addr_t;
+typedef uintptr_t vaddr_t;
+
+#define LTRACEF(...)
+#define LTRACE_ENTRY
+#define DEBUG_ASSERT assert
+#define ASSERT(x) do {} while(false)
+#define USE(x) ((void)(x))
+#define STATIC_ASSERT(condition)
+#define dprintf(...) fprintf(__VA_ARGS__)
```

Comment:
Should we consider compiling these out in non-debug modes?

Reply:
For now they are out because they are only called in tree-shaken functions.